packages/utils/telemetry-utils/src/logger.ts

```diff
@@ -784,32 +784,55 @@ function convertToBasePropertyTypeUntagged(
 
 export const tagData = <
 	T extends TelemetryDataTag,
-	V extends Record<string, TelemetryEventPropertyTypeExt>,
+	U extends TelemetryEventPropertyType,
+	V extends Record<string, U | (() => U)>,
 >(
 	tag: T,
 	values: V,
 ): {
 	[P in keyof V]:
 		| {
-				value: Exclude<V[P], undefined>;
+				value: Exclude<U, undefined>;
 				tag: T;
 		  }
+		| (() => {
+				value: Exclude<U, undefined>;
+				tag: T;
+		  })
 		| (V[P] extends undefined ? undefined : never);
 } =>
 	(Object.entries(values) as [keyof V, V[keyof V]][])
 		.filter((e): e is [keyof V, Exclude<V[keyof V], undefined>] => e[1] !== undefined)
 		// eslint-disable-next-line unicorn/no-array-reduce
 		.reduce<{
 			[P in keyof V]:
-				| (V[P] extends undefined ? undefined : never)
-				| { value: Exclude<V[P], undefined>; tag: T };
+				| {
+						value: Exclude<U, undefined>;
+						tag: T;
+				  }
+				| (() => {
+						value: Exclude<U, undefined>;
+						tag: T;
+				  })
+				| (V[P] extends undefined ? undefined : never);
 		}>((pv, cv) => {
-			pv[cv[0]] = { tag, value: cv[1] };
+			if (typeof cv[1] === "function") {
```

Comment:
do something like `const [key,value] = cv` this should help with type narrowing, and will make the code a bit easier to follow